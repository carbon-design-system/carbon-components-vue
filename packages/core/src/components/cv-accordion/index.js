import CvAccordion from './cv-accordion';
import CvAccordionItem from './cv-accordion-item';
import CvAccordionSkeleton from './cv-accordion-skeleton';

export { CvAccordion, CvAccordionItem, CvAccordionSkeleton };
