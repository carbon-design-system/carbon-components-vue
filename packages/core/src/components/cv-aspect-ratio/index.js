import CvAspectRatio from './cv-aspect-ratio';
export { CvAspectRatio };
