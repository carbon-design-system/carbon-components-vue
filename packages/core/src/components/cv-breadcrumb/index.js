import CvBreadcrumb from './cv-breadcrumb';
import CvBreadcrumbItem from './cv-breadcrumb-item';
import CvBreadcrumbSkeleton from './cv-breadcrumb-skeleton';

export { CvBreadcrumb, CvBreadcrumbItem, CvBreadcrumbSkeleton };
