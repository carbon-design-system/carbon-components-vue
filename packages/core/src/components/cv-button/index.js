import CvButton from './cv-button';
import CvIconButton from './cv-icon-button';
import CvButtonSkeleton from './cv-button-skeleton';
import CvButtonSet from './cv-button-set';

export { CvButton, CvIconButton, CvButtonSkeleton, CvButtonSet };
