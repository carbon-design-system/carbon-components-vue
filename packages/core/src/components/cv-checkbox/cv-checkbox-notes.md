# cv-checkbox

A Vue implementation of a Carbon Components checkbox

http://www.carbondesignsystem.com/components/checkbox/code

## Usage

### Default

```html
<cv-checkbox label="checkbox" @change="actionChange"> </cv-checkbox>
```

### mixed

```html
<cv-checkbox label="checkbox" mixed @change="actionChange"> </cv-checkbox>
```

### With v-model

```html
<cv-checkbox label="checkbox" v-model="checked" @change="actionChange"> </cv-checkbox>
```

```javascript
  data() {
    return {
      checked: Boolean,
    },
  }
  //...
```

### With v-model Array

```
html
<cv-checkbox v-model="checks" name="check-1" value="check-1" @change="actionChange" label="check-1"></cv-checkbox>
<cv-checkbox v-model="checks" name="check-2" value="check-2" @change="actionChange" label="check-2" mixed></cv-checkbox>
<cv-checkbox v-model="checks" name="check-3" value="check-3" @change="actionChange" label="check-3"></cv-checkbox>
<cv-checkbox v-model="checks" name="check-4" value="check-4" @change="actionChange" label="check-4" mixed></cv-checkbox>
```

```javascript
  data() {
    return {
      checks: ['check-3', 'check-4'], // check-3 and check-4 initially checked
    },
  }
  //...
```

## Attributes

- checked: true | false
- disabled: standard HTML checkbox property
- hideLabel: makes the label visually hidden but still labels the checkbox
- label: checkbox label
- mixed: if true aria-checkbox set to mixed if checked is false
