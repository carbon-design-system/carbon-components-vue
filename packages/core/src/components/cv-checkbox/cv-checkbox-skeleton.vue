<template>
  <div class="cv-checkbox" :class="`${carbonPrefix}--form-item ${carbonPrefix}--checkbox-wrapper`">
    <label :class="`${carbonPrefix}--checkbox-label ${carbonPrefix}--skeleton`">
      <span :class="`${carbonPrefix}--checkbox-label-text ${carbonPrefix}--skeleton`"></span>
    </label>
  </div>
</template>

<script>
import { carbonPrefixMixin } from '../../mixins';
export default {
  name: 'CvCheckboxSkeleton',
  mixins: [carbonPrefixMixin],
};
</script>
