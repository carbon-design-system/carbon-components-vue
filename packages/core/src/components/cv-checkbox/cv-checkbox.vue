<template>
  <div class="cv-checkbox" :class="[`${carbonPrefix}--checkbox-wrapper`, { [`${carbonPrefix}--form-item`]: formItem }]">
    <input
      ref="input"
      v-bind="$attrs"
      v-on="inputListeners"
      :class="`${carbonPrefix}--checkbox`"
      type="checkbox"
      :checked="isChecked === true"
      :aria-checked="`${isChecked}`"
      @focus="onFocus"
      @blur="onBlur"
      :value="value"
      :id="uid"
    />
    <label
      :class="[
        `${carbonPrefix}--checkbox-label`,
        {
          [`${carbonPrefix}--label--disabled`]: $attrs.disabled !== undefined && this.$attrs.disabled,
          [`${carbonPrefix}--checkbox-label__focus`]: hasFocus,
        },
      ]"
      :data-contained-checkbox-state="isChecked"
      :data-contained-checkbox-disabled="$attrs.disabled"
      :for="uid"
    >
      <span :class="[`${carbonPrefix}--checkbox-label-text`, { [`${carbonPrefix}--visually-hidden`]: hideLabel }]">
        {{ label }}
      </span>
    </label>
  </div>
</template>

<script>
import { checkMixin, uidMixin, carbonPrefixMixin, methodsMixin } from '../../mixins';

export default {
  name: 'CvCheckbox',
  mixins: [checkMixin, uidMixin, carbonPrefixMixin, methodsMixin({ input: ['blur', 'focus'] })],
  inheritAttrs: false,
  props: {
    hideLabel: Boolean,
    label: String,
    mixed: Boolean,
    formItem: { type: Boolean, default: true },
  },
  watch: {
    mixed() {
      this.dataMixed = this.mixed;
      if (this.dataMixed && this.checked !== true) {
        this.isChecked = false; // reset check state so mixed takes
      }
    },
  },
  data() {
    return {
      hasFocus: false,
      dataMixed: this.mixed,
    };
  },
  methods: {
    onFocus() {
      this.hasFocus = true;
    },
    onBlur() {
      this.hasFocus = false;
    },
  },
};
</script>
