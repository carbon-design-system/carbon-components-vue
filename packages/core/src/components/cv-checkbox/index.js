import CvCheckbox from './cv-checkbox';
import CvCheckboxSkeleton from './cv-checkbox-skeleton';
export { CvCheckbox, CvCheckboxSkeleton };
