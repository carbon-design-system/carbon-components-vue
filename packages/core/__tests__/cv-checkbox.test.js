// import { shallowMount as shallow } from '@vue/test-utils';
import { testComponent, awaitNextTick } from './_helpers';
const { shallowMount: shallow, trigger, setProps } = awaitNextTick;
import { CvCheckbox, CvCheckboxSkeleton } from '@/components/cv-checkbox';
// import { settings as carbonSettings } from 'carbon-components';

// const carbonPrefix = carbonSettings.prefix;

describe('CvCheckbox', () => {
  // ***************
  // PROP CHECKS
  // ***************
  testComponent.propsAreType(CvCheckbox, ['label', 'name', 'value'], String);
  testComponent.propsAreType(CvCheckbox, ['mixed', 'formItem', 'checked'], Boolean);
  testComponent.propsAreType(CvCheckbox, ['modelValue'], [Array, Boolean]);

  testComponent.propsHaveDefault(CvCheckbox, ['formItem']);
  testComponent.propsHaveDefaultOfUndefined(CvCheckbox, ['modelValue', 'checked']);

  // ***************
  // SNAPSHOT TESTS
  // ***************

  it('should render', async () => {
    const propsData = { formItem: true, value: 'check-1', id: '1' };
    const wrapper = await shallow(CvCheckbox, { propsData });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render when disabled', async () => {
    const propsData = { formItem: true, value: 'check-1', disabled: true, id: '1' };
    const wrapper = await shallow(CvCheckbox, { propsData });

    expect(wrapper.html()).toMatchSnapshot();
  });

  // ***************
  // FUNCTIONAL TESTS
  // ***************
  it('should be unchecked when mixed is true and checked is false', async () => {
    const propsData = { formItem: true, value: 'check-1', mixed: true, checked: false };
    const wrapper = await shallow(CvCheckbox, { propsData });
    expect(wrapper.find('input').element.checked).toEqual(false);
  });

  it('should be checked when mixed is true and checked is true', async () => {
    const propsData = { formItem: true, value: 'check-1', mixed: true, checked: true };
    const wrapper = await shallow(CvCheckbox, { propsData });
    expect(wrapper.find('input').element.checked).toEqual(true);
  });

  it('computed isChecked should be false when checked prop is false', async () => {
    const propsData = { formItem: true, value: 'check-1', checked: false };
    const wrapper = await shallow(CvCheckbox, { propsData });
    expect(wrapper.vm.isChecked).toEqual(false);
  });

  it('computed isChecked should be true when checked prop is true', async () => {
    const propsData = { formItem: true, value: 'check-1', checked: true };
    const wrapper = await shallow(CvCheckbox, { propsData });
    expect(wrapper.vm.isChecked).toEqual(true);
  });

  it('computed isChecked should be "mixed" when checked prop is false and mixed it true', async () => {
    const propsData = { formItem: true, value: 'check-1', mixed: true, checked: false };
    const wrapper = await shallow(CvCheckbox, { propsData });
    expect(wrapper.vm.isChecked).toEqual('mixed');
  });

  // Changes to checked property affect the computed value isChecked (wrapper.vm.isChecked)
  it('computed isChecked should be true when checked prop was changed to true', async () => {
    const propsData = { formItem: true, value: 'check-1', checked: false };
    const wrapper = await shallow(CvCheckbox, { propsData });
    expect(wrapper.vm.isChecked).toEqual(false);
    await setProps(wrapper, { checked: true });
    expect(wrapper.vm.isChecked).toEqual(true);
  });

  // Changes to checked property affect the computed value isChecked (wrapper.vm.isChecked)
  it('computed isChecked should be false when checked prop was changed to false', async () => {
    const propsData = { formItem: true, value: 'check-1', checked: true };
    const wrapper = await shallow(CvCheckbox, { propsData });
    expect(wrapper.vm.isChecked).toEqual(true);
    await setProps(wrapper, { checked: false });
    expect(wrapper.vm.isChecked).toEqual(false);
  });

  it('computed isChecked should be true when checked prop is true and mixed it true', async () => {
    const propsData = { formItem: true, value: 'check-1', mixed: true, checked: true };
    const wrapper = await shallow(CvCheckbox, { propsData });
    expect(wrapper.vm.isChecked).toEqual(true);
  });
  // Using modelValue property overrides checked.
  it('computed isChecked should be true when checked prop is false and modelValue is true', async () => {
    const propsData = { formItem: true, value: 'check-1', modelValue: true, checked: false };
    const wrapper = await shallow(CvCheckbox, { propsData });
    expect(wrapper.vm.isChecked).toEqual(true);
  });

  // Using modelValue property overrides checked.
  it('computed isChecked should be false when checked prop is true and modelValue is false', async () => {
    const propsData = { formItem: true, value: 'check-1', modelValue: false, checked: true };
    const wrapper = await shallow(CvCheckbox, { propsData });
    expect(wrapper.vm.isChecked).toEqual(false);
  });

  // Using an array with modelValue functions as expected.
  it('computed isChecked should be false when modelValue array is empty', async () => {
    const propsData = { formItem: true, value: 'check-1', modelValue: [] };
    const wrapper = await shallow(CvCheckbox, { propsData });
    expect(wrapper.vm.isChecked).toEqual(false);
  });

  // Using an array with modelValue functions as expected.
  it('computed isChecked should be false when modelValue array does not contain the value', async () => {
    const propsData = { formItem: true, value: 'check-1', modelValue: ['check-2', 'check-3'] };
    const wrapper = await shallow(CvCheckbox, { propsData });
    expect(wrapper.vm.isChecked).toEqual(false);
  });

  // Using an array with modelValue functions as expected.
  it('computed isChecked should be true when modelValue array contains the value', async () => {
    const propsData = { formItem: true, value: 'check-1', modelValue: ['check-2', 'check-3', 'check-1'] };
    const wrapper = await shallow(CvCheckbox, { propsData });
    expect(wrapper.vm.isChecked).toEqual(true);
  });

  it('input should have checkbox type', async () => {
    const propsData = { formItem: true, value: 'check-1' };
    const wrapper = await shallow(CvCheckbox, { propsData });
    expect(wrapper.find('input').element.type).toBe('checkbox');
  });

  it('should emit change event on click', async () => {
    const propsData = { formItem: true, value: 'check-1' };
    const wrapper = await shallow(CvCheckbox, { propsData });
    await trigger(wrapper.find('input'), 'click');
    expect(wrapper.emitted().change).toBeTruthy();
  });

  it('should emit modelEvent event on click', async () => {
    const propsData = { formItem: true, value: 'check-1' };
    const wrapper = await shallow(CvCheckbox, { propsData });
    await trigger(wrapper.find('input'), 'click');
    expect(wrapper.emitted().modelEvent).toBeTruthy();
  });

  // Clicking updates the value isChecked (wrapper.vm.isChecked)
  it('clicking updates the value isChecked from false to true', async () => {
    const propsData = { formItem: true, value: 'check-1', checked: false };
    const wrapper = await shallow(CvCheckbox, { propsData });
    await trigger(wrapper.find('input'), 'click');
    expect(wrapper.vm.isChecked).toEqual(true);
  });

  // Clicking updates the value isChecked (wrapper.vm.isChecked)
  it('clicking updates the value isChecked from true to false', async () => {
    const propsData = { formItem: true, value: 'check-1', checked: true };
    const wrapper = await shallow(CvCheckbox, { propsData });
    await trigger(wrapper.find('input'), 'click');
    expect(wrapper.vm.isChecked).toEqual(false);
  });
});

describe('CvCheckboxSkeleton', () => {
  // ***************
  // PROP CHECKS
  // ***************

  // ***************
  // SNAPSHOT TESTS
  // ***************
  it('Renders as expected', async () => {
    const wrapper = await shallow(CvCheckboxSkeleton);

    expect(wrapper.html()).toMatchSnapshot();
  });

  // ***************
  // FUNCTIONAL TESTS
  // ***************
});
