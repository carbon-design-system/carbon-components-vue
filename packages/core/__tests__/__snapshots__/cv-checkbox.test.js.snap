// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CvCheckbox should render 1`] = `
<div class="cv-checkbox bx--checkbox-wrapper bx--form-item"><input type="checkbox" aria-checked="false" id="1" class="bx--checkbox" value="check-1"> <label for="1" class="bx--checkbox-label"><span class="bx--checkbox-label-text">

    </span></label></div>
`;

exports[`CvCheckbox should render when disabled 1`] = `
<div class="cv-checkbox bx--checkbox-wrapper bx--form-item"><input type="checkbox" aria-checked="false" id="1" disabled="disabled" class="bx--checkbox" value="check-1"> <label data-contained-checkbox-disabled="true" for="1" class="bx--checkbox-label bx--label--disabled"><span class="bx--checkbox-label-text">

    </span></label></div>
`;

exports[`CvCheckboxSkeleton Renders as expected 1`] = `<div class="cv-checkbox bx--form-item bx--checkbox-wrapper"><label class="bx--checkbox-label bx--skeleton"><span class="bx--checkbox-label-text bx--skeleton"></span></label></div>`;
